src/python/pants/engine/target.py

```diff
@@ -1357,9 +1285,55 @@ def eager_fileset_with_spec(self, *, address: Address) -> EagerFilesetWithSpec:
         return EagerFilesetWithSpec(address.spec_path, self.filespec, self.snapshot)
 
 
+class CodegenSources(Sources):
+    pass
```

Comment:
This class allows us to very quickly determine if we should attempt to codegen or not on a target. If it only subclasses `Sources` and not `CodegenSources`, then we should not try to codegen.

I didn't add docstring because it would override what shows up in `./pants target-types --details`.. 🤷‍♂️

Reply:
> it is totally reasonable for a plugin author to want to convert JavaSources into EnrichedJavaSources

Huh, interesting. Could you share more about this use case please? Why use codegen rather than creating a new target type, like `enriched_java_library`? The Target API means that your `enriched_java_library` can be plugged in anywhere that `java_library` is used.

--

Agreed that in the `hydrate_sources` rule, we could remove this marker type to instead check if a codegen edge exists. 

Where this gets really clunky, though, is when we need to filter out irrelevant targets elsewhere. See https://github.com/pantsbuild/pants/pull/9638, where we only want `files()`, `resources()`, and Python targets. 

Soon, we will also want all codegen targets. This will soon be achieved with a simple `if tgt.has_field(CodegenSources)`.

Alternatively, if we remove this marker type, we would need to change the rule to take `UnionMembership` as a param, then get the `CodeGenerator` members and iterate over them to see if there's an edge or not. I'm sure we could factor this up, but even with factoring, it will be more complex than `if tgt.has_field(CodegenSources)`.